brainiak/utils/fmrisim.py

```diff
@@ -899,94 +1184,190 @@ def Pk2(idxs):
+                                          spatial_sigma=spatial_sigma,
+                                          )
 
-    return noise
+    volume_autoreg = _generate_noise_spatial(dimensions=dimensions,
+                                             mask=mask,
+                                             spatial_sigma=spatial_sigma,
+                                             )
+
+    # Multiply the noise by the spatial volume
+    noise_drift_volume = np.multiply.outer(volume_drift, noise_drift)
+    noise_phys_volume = np.multiply.outer(volume_phys, noise_phys)
+    noise_autoregression_volume = np.multiply.outer(volume_autoreg,
+                                                    noise_autoregression)
+
+    # Sum the noise (it would have been nice to just add all of them in a
+    # single line but this was causing formatting problems)
+    noise_temporal = noise_drift_volume * drift_sigma
+    noise_temporal = noise_temporal + (noise_phys_volume * physiological_sigma)
+    noise_temporal = noise_temporal + (noise_autoregression_volume *
+                                       auto_reg_sigma)
+
+    # Only do this if you are making motion variance
+    if motion_sigma != 0 and np.sum(stimfunction) > 0:
+        # Make each noise type
+        stimfunction_tr = stimfunction[::(tr_duration * 1000)]
+        noise_task = _generate_noise_temporal_task(stimfunction_tr,
+                                                   )
+        volume_task = _generate_noise_spatial(dimensions=dimensions,
+                                              mask=mask,
+                                              spatial_sigma=spatial_sigma,
+                                              )
+        noise_task_volume = np.multiply.outer(volume_task, noise_task)
+        noise_temporal = noise_temporal + (noise_task_volume * motion_sigma)
+
+    return noise_temporal
 
 
 def mask_brain(volume,
-               mask_name=None):
+               mask_name=None,
+               mask_threshold=0.1,
+               ):
     """ Mask the simulated volume
+    This takes in a volume and will output the masked volume. if a one
+    dimensional array is supplied then the output will be a volume of the
+    dimensions specified in the array. The mask can be created from the
+    volume by averaging it. All masks will be bounded to the range of 0 to 1.
 
     Parameters
     ----------
 
     volume : multidimensional array
-        The volume that has been simulated
+        Either numpy array of the volume that has been simulated or a tuple
+        describing the dimensions of the mask to be created
 
     mask_name : str
-        What is the path to the mask to be loaded?
+        What is the path to the mask to be loaded? If set to self then it
+        makes a mask from the volume supplied (by averaging across time
+        points and changing the range). If empty then it defaults to an
+        MNI152 grey matter mask.
```

Comment:
This gives a special meaning to the path "self". It would be better to introduce an extra boolean parameter, perhaps called `self_mask`.